modules/storage/mysql-single-server/main.bicep

```diff
@@ -0,0 +1,212 @@
+param sourceServerResourceId string = ''
+
+@description('Optional. Enable ssl enforcement or not when connect to server.')
+@allowed([
+  'Enabled'
+  'Disabled'
+])
+param sslEnforcement string = 'Enabled'
+
+@description('Optional. Auto grow of storage.')
+param storageAutogrow bool = true
+
+@description('Optional. The storage size of the server.')
+param storageSizeGB int = 32
+
+@description('Optional. The version of the MySQL server.')
+@allowed([
+  '5.6'
+  '5.7'
+  '8.0'
+])
+param version string = '8.0'
+
+@description('Array of role assignment objects that contain the "roleDefinitionIdOrName" and "principalId" to define RBAC role assignments on this resource. In the roleDefinitionIdOrName attribute, provide either the display name of the role definition, or its fully qualified ID in the following format: "/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11"')
+param roleAssignments array = []
+
+var varPrivateEndpoints = [for endpoint in privateEndpoints: {
+  name: '${mysqlServer.name}-${endpoint.name}'
+  privateLinkServiceId: mysqlServer.id
+  groupIds: [
+    endpoint.groupId
+  ]
+  subnetId: endpoint.subnetId
+  privateDnsZones: contains(endpoint, 'privateDnsZoneId') ? [
+    {
+      name: 'default'
+      zoneId: endpoint.privateDnsZoneId
+    }
+  ] : []
+  manualApprovalEnabled: contains(endpoint, 'manualApprovalEnabled') ? endpoint.manualApprovalEnabled : false
+}]
+
+resource mysqlServer 'Microsoft.DBforMySQL/servers@2017-12-01' = {
+  name: serverName
+  location: location
+  tags: tags
+  sku: {
+    name: skuName
+  }
+  identity: {
+    type: 'SystemAssigned'
+  }
+  properties: {
+    createMode: createMode
+    administratorLogin: administratorLogin
+    administratorLoginPassword: administratorLoginPassword
+    version: version
+    minimalTlsVersion: minimalTlsVersion
+    infrastructureEncryption: infrastructureEncryption
+    sslEnforcement: sslEnforcement
+    storageProfile: {
+      storageMB: storageSizeGB * 1024
+      backupRetentionDays: backupRetentionDays
+      geoRedundantBackup: geoRedundantBackup
+      storageAutogrow: createMode == 'Replica' ? null : (storageAutogrow ? 'Enabled' : 'Disabled')
+    }
+    publicNetworkAccess: publicNetworkAccess
+    sourceServerId: createMode != 'Default' ? sourceServerResourceId : json('null')
+    restorePointInTime: createMode == 'PointInTimeRestore' ? restorePointInTime : json('null')
+  }
+}
+
+@batchSize(1)
+resource mysqlServerFirewallRules 'Microsoft.DBforMySQL/servers/firewallRules@2017-12-01' = [for firewallRule in firewallRules: {
+  name: firewallRule.name
+  parent: mysqlServer
+  properties: {
+    startIpAddress: firewallRule.startIpAddress
+    endIpAddress: firewallRule.endIpAddress
+  }
+}]
```

Comment:
```suggestion
resource mysqlServer 'Microsoft.DBforMySQL/servers@2017-12-01' = {
  name: serverName
  location: location
  tags: tags
  sku: {
    name: skuName
  }
  identity: {
    type: 'SystemAssigned'
  }
  properties: {
    createMode: createMode
    administratorLogin: administratorLogin
    administratorLoginPassword: administratorLoginPassword
    version: version
    minimalTlsVersion: minimalTlsVersion
    infrastructureEncryption: infrastructureEncryption
    sslEnforcement: sslEnforcement
    storageProfile: {
      storageMB: storageSizeGB * 1024
      backupRetentionDays: backupRetentionDays
      geoRedundantBackup: geoRedundantBackup
      storageAutogrow: createMode == 'Replica' ? null : (storageAutogrow ? 'Enabled' : 'Disabled')
    }
    publicNetworkAccess: publicNetworkAccess
    sourceServerId: createMode != 'Default' ? sourceServerResourceId : json('null')
    restorePointInTime: createMode == 'PointInTimeRestore' ? restorePointInTime : json('null')
  }

  @batchSize(1)
  resource mysqlServerFirewallRules 'firewallRules' = [for firewallRule in firewallRules: {
    name: firewallRule.name
    properties: {
      startIpAddress: firewallRule.startIpAddress
      endIpAddress: firewallRule.endIpAddress
    }
  }]
}
```

Reply:
@dciborow I tried your suggestion, 
https://learn.microsoft.com/en-us/azure/azure-resource-manager/bicep/child-resource-name-type

But @batch is not supported inside the parent resource and for that reason we are following this approach:
https://learn.microsoft.com/en-us/azure/azure-resource-manager/bicep/child-resource-name-type

Reply:
ah, in that case, it has to be how it is. No worries then.